moveit_planners/trajopt/src/trajopt_interface.cpp

```diff
@@ -0,0 +1,458 @@
+
+namespace trajopt_interface
+{
+TrajOptInterface::TrajOptInterface(const ros::NodeHandle& nh) : nh_(nh), name_("TrajOptInterface")
+{
+  prob_ = TrajOptProblemPtr(new TrajOptProblem);
+  setDefaultTrajOPtParams();
+
+  // TODO: callbacks should be defined by the user
+  callbacks_.push_back(callBackFunc);
+}
+
+bool TrajOptInterface::solve(const planning_scene::PlanningSceneConstPtr& planning_scene,
+                             const planning_interface::MotionPlanRequest& req,
+                             moveit_msgs::MotionPlanDetailedResponse& res)
+{
+  ROS_INFO(" ======================================= From trajopt_interface, solve is called");
+  setTrajOptParams(params_);
+
+  if (!planning_scene)
+  {
+    ROS_ERROR_STREAM_NAMED("trajopt_planner", "No planning scene initialized.");
+    res.error_code.val = moveit_msgs::MoveItErrorCodes::FAILURE;
+    return false;
+  }
+
+  if (req.start_state.joint_state.position.empty())
+  {
+    ROS_ERROR_STREAM_NAMED("trajopt_planner", "Start state position is empty");
+    res.error_code.val = moveit_msgs::MoveItErrorCodes::INVALID_ROBOT_STATE;
+    return false;
+  }
+
+  if (req.start_state.joint_state.name.empty())
+  {
+    ROS_ERROR_STREAM_NAMED("trajopt_planner", "Start_state name is empty");
+    res.error_code.val = moveit_msgs::MoveItErrorCodes::INVALID_ROBOT_STATE;
+    return false;
+  }
+
+  ROS_INFO(" ======================================= Extract current state information");
+  ros::WallTime start_time = ros::WallTime::now();
+  robot_model::RobotModelConstPtr robot_model = planning_scene->getRobotModel();
+  robot_state::RobotStatePtr current_state(new robot_state::RobotState(robot_model));
+  *current_state = planning_scene->getCurrentState();
+  const robot_state::JointModelGroup* joint_model_group = current_state->getJointModelGroup(req.group_name);
+  std::vector<std::string> joint_names = joint_model_group->getVariableNames();
+  int dof = joint_names.size();
+  std::vector<double> current_joint_values;
+  current_state->copyJointGroupPositions(joint_model_group, current_joint_values);
+
+  // current state is different from star state in general
+  ROS_INFO(" ======================================= Extract start stae infromation");
+  std::unordered_map<std::string, double> all_joints;
+  trajopt::DblVec joint_start_constraints;
+
+  for (int joint_index = 0; joint_index < req.start_state.joint_state.position.size(); ++joint_index)
+  {
+    all_joints[req.start_state.joint_state.name[joint_index]] = req.start_state.joint_state.position[joint_index];
+  }
+
+  for (auto joint_name : joint_names)
+  {
+    ROS_INFO(" joint position from start state, name: %s, value: %f", joint_name.c_str(), all_joints[joint_name]);
+    joint_start_constraints.push_back(all_joints[joint_name]);
+  }
+
+  // check the joint limit violiation for star state
+  if (not planning_scene->getRobotModel()->satisfiesPositionBounds(joint_start_constraints.data()))
```

Comment:
The position bounds should be checked using the joint model group or otherwise the number and order of joint values could be invalid. Could you change this to use `JointModelGroup::satisfiesPositionBounds()` and maybe verify that the order of the joint values matches the order specified in the group?

Reply:
used joint_model_group. 
joint values in joint_start_constraints correspond to the ones in the joint model group.